### ISSUE TITLE
PPO2: act_model and train_model[question]

### ISSUE THREAD

**Issue by denyHell** [hill-a/stable-baselines#860]
This is a question on the implementation of PPO2 in stable baselines. I notice that there are two models: the act model and the train model. Am I correct that they share the same weights, and the only difference is the input dimensional? If that is the case, what's the difference between
`vpred = train_model.value_flat` and  `self.old_vpred_ph` , which is fed the act_model.value_flat during training? 
I am also a bit confused on the lines
`  vpred_clipped = self.old_vpred_ph + \
                        tf.clip_by_value(train_model.value_flat - self.old_vpred_ph,
                                             - self.clip_range_vf_ph, self.clip_range_vf_ph)`
Why doing this and why it depends on reward scaling?

Thanks in advance!

**Comment by Miffyli** [comment]
Similar to how policy is updated by clipping the fraction between current and old (one that gathered samples) policy, this code here does the same for values (i.e. avoid too large changes from old policy). `self.old_vpred_ph` includes these original predictions, and the new ones after updates should not deviate _too_ much from them. `vpred_clipped` is this clip value.

This feature was not explained in the PPO paper, but is present in [the baselines implementation](https://github.com/openai/baselines/blob/master/baselines/ppo2/model.py#L69).

**Comment by denyHell** [comment]
i guess where i got confused is (1) are train model and act model the same? (2) if not, where in the code do we sync act model to the train model?

**Comment by Miffyli** [comment]
Ah yes. I agree this part is confusing and I do not fully understand it myself, but indeed the variables seem to be shared and only difference is the number of outputs. All the "values during acting" are stored (and frozen) in the placeholders. The naming here is bit confusing.

@araffin Do you have insight on what is the purpose of these, exactly? For training LSTMs?

**Comment by denyHell** [comment]
It just does not make sense if act model and train model are the same...act model is the old policy, while train model is the new one, and the clipping in objective function is to make sure that they do not differ too much.

**Comment by araffin** [comment]
>(1) are train model and act model the same?

yes. `act_model` has nothing to do with training, it is only used at predict time.
The question was already answered in https://github.com/hill-a/stable-baselines/issues/222 and https://github.com/hill-a/stable-baselines/issues/314

And @Miffyli is right, we have two because of tensorflow not handling properly different input shapes for recurrent policies.

>vpred_clipped

This is an un-documented trick from the Baselines code. It enforces that value prediction is not too far from the old one, but it depends on the reward magnitude (see #216 )

There is a paper about the different code-level optimizations and their effects here: https://openreview.net/forum?id=r1etN1rtPB

>clipping in objective function is to make sure that they do not differ too much.

The clipping is done differently, you don't need to copy the weights, you just need to keep track of the old log probabilities:
https://github.com/hill-a/stable-baselines/blob/c4c31cb5687800cf102054f0935aa09e7545159f/stable_baselines/ppo2/ppo2.py#L142

"old negative log prob actions" (yes, quite ugly name... legacy code...)
I recommend you to read the v3 version ;) (much cleaner)
https://github.com/DLR-RM/stable-baselines3/blob/master/stable_baselines3/ppo/ppo.py#L198

the same tricks (objective and value clipping) are implemented by in pytorch.

**Comment by denyHell** [comment]
@araffin Thanks for your reply. 

(1) It seems that `self.old_neglog_pac_ph` will be fed the values of negative log prob actions computed using the act model (i.e. `self.model.step` in the function runner.run()). On the other hand, the new value `neglogpac = train_model.proba_distribution.neglogp(self.action_ph)` is the exact same value computed using train model. If these two models are the same, then why are these two values different (so that the ratio ` ratio = tf.exp(self.old_neglog_pac_ph - neglogpac)` is not always 1)?
(2) From the line `vpred_clipped = self.old_vpred_ph + tf.clip_by_value(train_model.value_flat - self.old_vpred_ph, - self.clip_range_vf_ph, self.clip_range_vf_ph)` , we again see that `self.old_vpred_ph` which is the value function of act model, and `train_model.value_flat` are different. Wouldn't that imply that 

> value function of act model = $V_{\theta_{t-1}}$
> value function of train model = $V_{\theta_t}$ ?

i.e. train model uses the updated weights while act model uses the weight 1 step before?

**Comment by araffin** [comment]
the `ratio = tf.exp(self.old_neglog_pac_ph - neglogpac)` will be one at the first iteration, but afterward, the values fed into `old_neglog_pac_ph` are the same (we don't use the act model to compute them again, otherwise, the ratio will be always one).

the answer for the 2nd question is the same: we use `act_model` (same as `train_model`) only before the first epoch to compute those values, that stay the same during the rest of the optimization steps (we don't re-compute them).

again, I really recommend you to read v3 code, the logic is exactly the same, but in pytorch (so closer to python), because I think the problem here is more because of tensorflow logic that is harder to read.

**Comment by denyHell** [comment]
Understood, will definitely take a look. Thanks a lot !

**Comment by Kailiangdong** [comment]
Hello, I have an additional question. I saw n_step = 1 in act_model and n_step=64 or 128 in train_model. The n_step means how many steps of data are taken as time-series data for LSTM.   But it seems that there is a fully connected layer after lstm layer for action distribution generation. How could these two models with different n_steps(also I mean different dimensions for fully connect layer and different numbers of weights) share parameters?

Thank you very much.

**Comment by Miffyli** [comment]
@Kailiangdong 

LSTM removes the sequence-dimension, essentially turning the 64/128 `n_step` sequence into output of same size as with 1 `n_step`, so the same fully connected network works in both cases.
